src/parser/jobs/blm/modules/RotationWatchdog.tsx

```diff
@@ -0,0 +1,563 @@
+	EXTRA_F1: 50, // These two codes should stay close to each other
+	NO_FIRE_SPELLS: 75,
+	DROPPED_ENOCHIAN: 100,
+}
+
+/**
+ * If you add an error code above, give it a message here too, even ones filtered out normally
+ * so they show something in debug mode.
+ */
+const ERROR_MESSAGES = {
+	[ERROR_CODES.NONE]: <Trans id="blm.rotation-watchdog.error-messages.none">No errors</Trans>,
+	[ERROR_CODES.FINAL_OR_DOWNTIME]: <Trans id="blm.rotation-watchdog.error-messages.final-or-downtime">Ended with downtime, or last cycle</Trans>,
+	[ERROR_CODES.SHORT]: <Trans id="blm.rotation-watchdog.error-messages.short-rotation">Too short, won't process</Trans>,
+	[ERROR_CODES.DROPPED_ENOCHIAN]: <Trans id="blm.rotation-watchdog.error-messages.dropped-enochian">Dropped <ActionLink {...ACTIONS.ENOCHIAN}/></Trans>,
+	[ERROR_CODES.MISSING_FIRE4S]: <Trans id="blm.rotation-watchdog.error-messages.missing-fire4s">Missing <ActionLink {...ACTIONS.FIRE_IV}/>s</Trans>,
+	[ERROR_CODES.EXTRA_T3]: <Trans id="blm.rotation-watchdog.error-messages.extra-t3">Extra <ActionLink {...ACTIONS.THUNDER_III}/>s</Trans>,
+	[ERROR_CODES.MANAFONT_BEFORE_DESPAIR]: <Trans id="blm.rotation-watchdog.error-messages.manafont-before-despair"><ActionLink {...ACTIONS.MANAFONT}/> used before <ActionLink {...ACTIONS.DESPAIR}/></Trans>,
+	[ERROR_CODES.MISSING_DESPAIRS]: <Trans id="blm.rotation-watchdog.error-messages.missing-despair">Rotation didn't include <ActionLink {...ACTIONS.DESPAIR}/></Trans>,
+	[ERROR_CODES.NO_FIRE_SPELLS]: <Trans id="blm.rotation-watchdog.error-messages.no-fire-spells">Rotation included no Fire spells</Trans>,
+	[ERROR_CODES.EXTRA_F1]: <Trans id="blm.rotation-watchdog.error-messages.extra-f1">Extra <ActionLink {...ACTIONS.FIRE_I}/></Trans>,
+}
+
+class Cycle {
+	casts: any[] = []
+	startTime: number
+	endTime?: number
+
+	hasManafont: boolean = false
+
+	inFirePhase: boolean = false
+	atypicalAFStart: boolean = false
+	firePhaseStartMP: number = 0
+
+	firstCycle: boolean = false
+	finalOrDowntime: boolean = false
+
+	gaugeStateBeforeFire: GaugeState = new GaugeState()
+
+	_errorCode: number = 0
+	public set errorCode(code: number) {
+		this._errorCode = Math.max(this._errorCode, code)
+	}
+	public get errorCode() {
+		return this._errorCode
+	}
+
+	/**
+	 * Greatly simplified from the pre-Shadowbringers version of this function. Because Umbral Soul gives us
+	 * a proper downtime action to build UH/UI, it should be possible to enter every fire phase with a normal
+	 * gauge state, so we're no longer going to reduce the expected count based on the actual gauge state on
+	 * entering fire phase. We're still tracking the data necessary to do this again in the future if there is
+	 * value in doing so.
+	 */
+	public get expectedFire4s(): number | undefined {
+		if (this.finalOrDowntime) {
+			return
+		}
+		// Account for the no-UH opener when determining the expected count of Fire 4s
+		return (this.firstCycle && this.gaugeStateBeforeFire.umbralHearts === 0 ?
```

Comment:
this is getting a bit long - if it gets much longer, consider splitting each partial calc into it's own var that can be added together.

Reply:
Midas whale do it now